File: cutOffTreesForGolfEvent/cutOffTreesForGolfEvent.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    class Compare {
    public:
        bool operator()(const tuple<int, int, int>& t1, const tuple<int, int, int>& t2) const {
            return get<0>(t1) < get<0>(t2);
        }
    };
    int cutOffTree(vector<vector<int>>& forest) {
        if (forest.size() == 0 || forest[0].size() == 0)
            return -1;
        int m = forest.size(), n = forest[0].size();
        vector<tuple<int, int, int> > trees;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (forest[i][j] != 0) {
                    trees.emplace_back(forest[i][j], i, j);
                }
            }
        }
        if (trees.size() == 0)
            return -1;
        sort(trees.begin(), trees.end(), Compare());
        vector<pair<int, int> > dirs = {{-1,0}, {1,0}, {0,-1}, {0, 1}};
        int res = bfs(forest, 0, 0, get<1>(trees[0]), get<2>(trees[0]), dirs);
        if (res == -1)
            return res;
        for (int i = 1; i < trees.size(); i++) {
            int dist = bfs(forest, get<1>(trees[i-1]), get<2>(trees[i-1]), get<1>(trees[i]), get<2>(trees[i]), dirs);
            if (dist == -1)
                return -1;
            else
                res += dist;
        }
        return res;
    }
    int bfs(vector<vector<int> >& forest, int si, int sj, int di, int dj, vector<pair<int, int> >& dirs) {
        int dist = 0;
        if (si == di && sj == dj)
            return dist;
        int m = forest.size(), n = forest[0].size();
        vector<vector<bool> > visited(m, vector<bool>(n, false));
        queue<pair<int, int> > q;
        q.emplace(si, sj);
        visited[si][sj] = true;
        
        while (!q.empty()) {
            int size = q.size();
            dist++;
            for (int i = 0; i < size; i++) {
                pair<int, int>& p = q.front();
                q.pop();
                for (pair<int, int>& dir : dirs) {
                    int next_i = p.first + dir.first;
                    int next_j = p.second + dir.second;
                    if (next_i < 0 || next_i >= m || next_j < 0 || next_j >= n || forest[next_i][next_j] == 0 || visited[next_i][next_j])
                        continue;
                    if (next_i == di && next_j == dj) {
                        return dist;
                    }
                    q.emplace(next_i, next_j);
                    visited[next_i][next_j] = true;
                }
            }
        }
        return -1;
    }
};

int main()
{
    Solution sl;
    vector<int> v1 = {1,2,3};
    vector<int> v2 = {0,0,4};
    vector<int> v3 = {7,6,5};
    vector<vector<int> > forest;
    forest.push_back(v1);
    forest.push_back(v2);
    forest.push_back(v3);
    cout << sl.cutOffTree(forest) << endl;
    return 0;
}

File: cutOffTreesForGolfEvent/cutOffTreesForGolfEvent.java
class Solution {
    public int cutOffTree(List<List<Integer>> forest) {
        int m = forest.size(), n = forest.get(0).size();
        List<int[]> path = new ArrayList<>();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (forest.get(i).get(j) > 1) {
                    path.add(new int[] {forest.get(i).get(j), i, j});
                }
            }
        }
        if (path.size() == 0) return 0;
        path.sort((a, b) -> a[0] - b[0]);
        
        int step = bfs(0, 0, path.get(0)[1], path.get(0)[2], forest, m, n);
        if (step == -1) return -1;
        int res = step;
        for (int i = 0; i < path.size() - 1; i++) {
            step = bfs(path.get(i)[1], path.get(i)[2], path.get(i + 1)[1], path.get(i + 1)[2], forest, m, n);
            if (step == -1) return -1;
            res += step;
        }
        return res;
    }
    
    private int bfs(int start_i, int start_j, int end_i, int end_j, List<List<Integer>> forest, int m, int n) {
        int res = 0;
        if (start_i == end_i && start_j == end_j) return res;
        boolean[][] visited = new boolean[m][n];
        int[][] dirs = {{-1, 0}, {1, 0}, {0, 1}, {0, -1}};
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[] {start_i, start_j});
        visited[start_i][start_j] = true;
        while (!q.isEmpty()) {
            int q_size = q.size();
            res++;
            for (int i = 0; i < q_size; i++) {
                int[] cur = q.poll();
                for (int[] dir : dirs) {
                    int next_i = cur[0] + dir[0];
                    int next_j = cur[1] + dir[1];
                    if (next_i < 0 || next_i >= m || next_j < 0 || next_j >= n) continue;
                    if (next_i == end_i && next_j == end_j) return res;
                    if (visited[next_i][next_j] || forest.get(next_i).get(next_j) == 0) continue;
                    q.offer(new int[]{next_i, next_j});
                    visited[next_i][next_j] = true;
                }
            }
        }
        return -1;
    }
}
